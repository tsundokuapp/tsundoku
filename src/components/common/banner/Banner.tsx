import Image from 'next/image';
import Link from 'next/link';
import type { ReactNode } from 'react';

import { merge } from '@/helpers/twUtils';

interface BannerProps {
  className?: string;
  children?: ReactNode;
  image: string;
  link?: string;
}

export function Banner({
  children,
  className,
  image,
  link,
  ...props
}: BannerProps) {
  return (
    <Link href={link || '#'}>
      <div
        className={merge(
          'flex h-72 items-center justify-center overflow-hidden rounded-md bg-appBannerBackground text-appBannerText',
          className,
        )}
        {...props}
      >
        <Image
          src={image}
          height={400}
          width={400}
          alt="Recrutamento de editores"
        />
        {children}
      </div>
    </Link>
  );
}
