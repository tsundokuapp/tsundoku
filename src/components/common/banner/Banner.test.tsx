import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';

import { Banner } from './Banner';

describe('Testes para o componente <Banner />', () => {
  describe('Testes de renderização', () => {
    it('Deve renderizar Banner corretamente', () => {
      // Given
      render(<Banner image="/recrutamento-editor.png" link="/example" />);

      // When
      const link = screen.getByRole('link');
      const image = screen.getByAltText('Recrutamento de editores');

      // Then
      expect(link).toHaveAttribute('href', '/example');
      expect(image).toBeInTheDocument();
      expect(image).toHaveAttribute(
        'src',
        expect.stringMatching(
          /\/_next\/image\?url=.*%2Frecrutamento-editor\.png/,
        ),
      );
    });

    it('Deve renderizar o conteúdo fornecido', () => {
      // Given
      render(
        <Banner image="/recrutamento-editor.png" link="/example">
          <p>Conteúdo personalizado</p>
        </Banner>,
      );

      // Then
      expect(screen.getByText('Conteúdo personalizado')).toBeInTheDocument();
    });

    it('Deve aplicar classes adicionais passadas via prop className', () => {
      // Given
      render(
        <Banner image="/recrutamento-editor.png" className="bg-red-500" />,
      );

      // When
      const banner = screen.getByRole('link').firstChild;

      // Then
      expect(banner).toHaveClass('bg-red-500');
    });
  });

  describe('Testes de interação e comportamento', () => {
    it('Deve usar "#" como link padrão se o link não for fornecido', () => {
      // Given
      render(<Banner image="/recrutamento-editor.png" />);

      // When
      const link = screen.getByRole('link');

      // Then
      expect(link).toHaveAttribute('href', '#');
    });

    it('Deve redirecionar para o link correto ao clicar', async () => {
      // Given
      render(<Banner image="/recrutamento-editor.png" link="/redirect" />);

      // When
      const link = screen.getByRole('link');
      await fireEvent.click(link);

      // Then
      expect(link).toHaveAttribute('href', '/redirect');
    });
  });

  describe('Teste de snapshot', () => {
    it('Deve manter o snapshot do Banner', () => {
      // Given
      const { container } = render(
        <Banner image="/recrutamento-editor.png" link="/example" />,
      );

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
