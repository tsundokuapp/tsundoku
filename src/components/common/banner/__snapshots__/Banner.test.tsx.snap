// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testes para o componente <Banner /> Teste de snapshot Deve manter o snapshot do Banner 1`] = `
<div>
  <a
    href="/example"
  >
    <div
      class="flex h-72 items-center justify-center overflow-hidden rounded-md bg-appBannerBackground text-appBannerText"
    >
      <img
        alt="Recrutamento de editores"
        data-nimg="1"
        decoding="async"
        height="400"
        loading="lazy"
        src="/_next/image?url=%2Frecrutamento-editor.png&w=828&q=75"
        srcset="/_next/image?url=%2Frecrutamento-editor.png&w=640&q=75 1x, /_next/image?url=%2Frecrutamento-editor.png&w=828&q=75 2x"
        style="color: transparent;"
        width="400"
      />
    </div>
  </a>
</div>
`;
